### PR TITLE
[Publisher] Group non-calendar year records under the year that matches their end year

### LINKED ISSUE #1141: [Publisher + Agency Dashboards] Adjust all charts to display non-calendar year records by end year instead of start year
A while back, we made changes to the period of time (year) a fiscal year record belongs to. Originally, a 2022-2023 fiscal year record (for example) would belong to the 2022 year - but, after the change, it belongs to 2023. At the time, we didn't quite realize this would require updates to:
* All of the FY data viz charts
* All of the FY records on the `/records` page

We should adjust the data viz charts and the records on the `/records` page to categorize non-calendar year records and datapoints by their end year.

### PR DESCRIPTION
**Pull request by mxosman**
## Description of the change

Groups non-calendar year annual records under the year that matches their end year.

Since we've adjusted the reference year for non-calendar year records/metrics to be their ending year, this is the last collateral adjustment that needs to happen to honor this in the UI. Currently, the records overview page takes a list of records, sorted by their year, and renders them in a linear way dropping report year markers in the UI as the list of records transitions from year to year.

In order to group non-calendar year records under their respective ending year, we need to adjust the structure of the records list and the rendering structure to accommodate this. If we group the list of records by their year - and while doing so, bucket non-calendar year records to their respective end year bucket:
```
{
  2024: [<Jan 2024 Record>, <Feb 2023-Jan 2024 Record>, <April 2023-March 2024 Record>],
  2023: [<Jan 2023 Record>, <Feb 2023 Record>, <Jul 2022-Jun 2023 Record>],
}
```
...we can then iterate through this object's entries and render the list of records by year respecting the end year of the non-calendar year records.

https://github.com/Recidiviz/justice-counts/assets/59492998/cbff8aeb-3268-4736-8744-e2d9c3e20c7e

## Related issues

Closes #1141 

### Development

**This box MUST be checked by the submitter prior to merging**:
- [ ] **Double- and triple-checked that there is no Personally Identifiable Information (PII) being mistakenly added in this pull request**

These boxes should be checked by the submitter prior to merging:
- [ ] Tests have been written to cover the code changed/added as part of this pull request

### Code review

These boxes should be checked by reviewers prior to merging:

- [ ] This pull request has a descriptive title and information useful to a reviewer
- [ ] This pull request has been moved out of a Draft state, has no "Work In Progress" label, and has assigned reviewers
- [ ] Potential security implications or infrastructural changes have been considered, if relevant

### CODE REVIEW

File: publisher/src/pages/Reports.tsx [hunk 1]
```diff
@@ -290,138 +310,150 @@ const Reports: React.FC = () => {
...
+                              isRecordDisabledForSelection(
+                                report.status,
+                                bulkAction
+                              )
+                            );
+                          }
+                        }}
+                        selected={
+                          bulkAction && selectedRecords.includes(report.id)
+                        }
```
Comment (mxosman) [inline publisher/src/pages/Reports.tsx:346]: No changes in this block.

File: publisher/src/pages/Reports.tsx [hunk 2]
```diff
@@ -290,138 +310,150 @@ const Reports: React.FC = () => {
...
+                          {bulkAction &&
+                            !isRecordDisabledForSelection(
+                              report.status,
+                              bulkAction
+                            ) && (
+                              <>
+                                {selectedRecords.includes(report.id) ? (
+                                  <SelectedCheckmark
+                                    src={checkmarkIcon}
+                                    alt=""
```
Comment (mxosman) [inline publisher/src/pages/Reports.tsx:359]: No changes in this block.

File: publisher/src/pages/Reports.tsx [hunk 3]
```diff
@@ -290,138 +310,150 @@ const Reports: React.FC = () => {
...
                           )}
-                    </Cell>
-                  </Row>
-
-                  {/* Report Year Marker */}
-                  {renderReportYearRow(
-                    filteredReportsMemoized,
-                    index,
-                    report.year
-                  )}
```
Comment (mxosman) [inline publisher/src/pages/Reports.tsx:420]: This was the only thing removed from this section - there's logic for the report year marker at the beginning of this render. We also no longer need this function.

File: publisher/src/pages/Reports.tsx [hunk 4]
```diff
@@ -290,138 +310,150 @@ const Reports: React.FC = () => {
...
+                                role={filteredReportEditors[0].role}
+                              />
+                              {filteredReportEditors.length > 1 ? (
+                                <AndOthersSpan>{`& ${
+                                  filteredReportEditors.length - 1
+                                } other${
+                                  filteredReportEditors.length > 2 ? "s" : ""
+                                }`}</AndOthersSpan>
+                              ) : null}
+                            </EditorsContentCellContainer>
```
Comment (mxosman) [inline publisher/src/pages/Reports.tsx:411]: No changes in this block.

File: publisher/src/pages/Reports.tsx [hunk 5]
```diff
@@ -290,138 +310,150 @@ const Reports: React.FC = () => {
...
-          )
+
+                          {filteredReportEditors.length > 1 && (
+                            <Tooltip
+                              anchorId={`record-${report.id}`}
+                              position="bottom"
+                              content={
+                                <EditorsTooltipContainer>
+                                  {filteredReportEditors.map((editor, idx) => (
+                                    <React.Fragment key={editor.name}>
```
Comment (mxosman) [inline publisher/src/pages/Reports.tsx:421]: No changes in this block.

File: publisher/src/pages/Reports.tsx [hunk 6]
```diff
@@ -290,138 +310,150 @@ const Reports: React.FC = () => {
...
+                            : printElapsedDaysMonthsYearsSinceDate(
+                                report.last_modified_at
+                              )}
+                        </Cell>
+                      </Row>
+                    </Fragment>
+                  );
+                })}
+            </>
+          ))
```
Comment (mxosman) [inline publisher/src/pages/Reports.tsx:456]: No changes in this block aside from removing the outlined area in this comment: https://github.com/Recidiviz/justice-counts/pull/1143/files#r1454194803

File: publisher/src/pages/Reports.tsx [hunk 7]
```diff
@@ -290,138 +310,150 @@ const Reports: React.FC = () => {
...
+              {new Date().getUTCFullYear() !== Number(reportYear) && (
+                <Row noHover isRowReportYear>
+                  {reportYear}
+                </Row>
+              )}
+
+              {reports
+                .sort(sortReportsAnnualFirstDescending)
+                .map((report: ReportOverview) => {
+                  const filteredReportEditors = filterJCAdminEditors(
```
Comment (mxosman) [inline publisher/src/pages/Reports.tsx:325]: This block has new changes.

File: publisher/src/pages/Reports.tsx [hunk 8]
```diff
@@ -290,138 +310,150 @@ const Reports: React.FC = () => {
...
+              {/* Only show report year marker for prior years (before current year) */}
+              {new Date().getUTCFullYear() !== Number(reportYear) && (
+                <Row noHover isRowReportYear>
+                  {reportYear}
+                </Row>
+              )}
+
+              {reports
+                .sort(sortReportsAnnualFirstDescending)
+                .map((report: ReportOverview) => {
```
Comment (mxosman) [inline publisher/src/pages/Reports.tsx:324]: This block has new changes.

File: publisher/src/pages/Reports.tsx [hunk 9]
```diff
@@ -161,25 +162,6 @@ const Reports: React.FC = () => {
...
-      indexIsLessThanListOfReports && filteredReports[currentIndex + 1].year;
-
-    if (indexIsLessThanListOfReports && nextReportYear !== currentReportYear) {
-      return (
-        <Row noHover isRowReportYear>
-          {nextReportYear}
-        </Row>
-      );
-    }
-  };
```
Comment (mxosman) [inline publisher/src/pages/Reports.tsx:181]: I don't know wth we were doing with this one, but thank goodness we don't need it anymore.

**Comment by lilidworkin** [comment]
@morden35 could you review this one too? Thank you!

**Comment by morden35** [comment]
> @morden35 could you review this one too? Thank you!

yes of course (sorry for the delay)! will review both today

File: publisher/src/pages/Reports.tsx (hunk 9 above)
Comment (morden35) [reply publisher/src/pages/Reports.tsx:181]: Looks like it was only used in line 416 which has now been removed as well

File: publisher/src/pages/Reports.tsx [hunk 10]
```diff
@@ -219,6 +201,24 @@ const Reports: React.FC = () => {
...
     [reportStore.reportOverviewList, reportsFilter]
   );
+  /**
+   * Groups the filtered reports by report year and adds non-calendar year annual reports to the
+   * year matching the report's ending year period (e.g. Feb 2023 - Jan 2024 report will belong to year 2024)
+   */
+  const filteredReportsMemoizedByReportYears = groupBy(
+    filteredReportsMemoized,
+    (report) =>
+      report.frequency === "ANNUAL" && report.month !== 1 // Note: `report.month` is not zero-indexed
```
Comment (morden35) [inline publisher/src/pages/Reports.tsx:211]: lovely comments again!

File: publisher/src/pages/Reports.tsx [hunk 11]
```diff
@@ -273,6 +273,26 @@ const Reports: React.FC = () => {
     highlight: ReportStatusFilterOptionObject[reportsFilter] === value,
   }));
 
+  const sortReportsAnnualFirstDescending = (
```
Comment (morden35) [inline publisher/src/pages/Reports.tsx:276]: woah! This is a very cool function but could maybe benefit from a docstring or some more comments

**Comment by morden35** [comment]
This is looking great! Would it be too much to also sort by month after we sort by year? For example, we can see in the screenshot below that if an agency has multiple different fiscal year reports, the months are not ordered. Preferably these reports would also be ordered in descending order. I think this is probably fine to do in a follow up pr if it's a heavy lift.

Screenshot 2024-01-18 at 10 53 35 AM

One idea to implement this would be to iterate on the structure below and turn the dictionary of lists into a dictionary of lists of dictionaries of lists:
```
{
  2024: [<Jan 2024 Record>, <Feb 2023-Jan 2024 Record>, <April 2023-March 2024 Record>],
  2023: [<Jan 2023 Record>, <Feb 2023 Record>, <Jul 2022-Jun 2023 Record>],
}
```
to
```
{
  2024: [
     {1: [<Jan 2024 Record>]},
     {2: [<Feb 2023-Jan 2024 Record>]},
     {4: [<April 2023-March 2024 Record>]}
     ],
  2023: [
     {1: [<Jan 2023 Record>]},
     {2: [<Feb 2023 Record>]},
     {7: [<Jul 2022-Jun 2023 Record>]},
     ]
}

lmk if that makes sense!

File: publisher/src/pages/Reports.tsx (hunk 9 above)
Comment (mxosman) [reply publisher/src/pages/Reports.tsx:181]: This was lovely to clean up. Such an intensely complex function past self (21 months ago?) wrote just to render the report year marker... Haha - wow, we were trying our best! 😆

File: publisher/src/pages/Reports.tsx (hunk 11 above)
Comment (mxosman) [reply publisher/src/pages/Reports.tsx:276]: Good call!

**Comment by mxosman** [comment]
Thank you so much for the review and feedback, Michelle!

> This is looking great! Would it be too much to also sort by month after we sort by year? For example, we can see in the screenshot below that if an agency has multiple different fiscal year reports, the months are not ordered. Preferably these reports would also be ordered in descending order. I think this is probably fine to do in a follow up pr if it's a heavy lift.
> 
> Screenshot 2024-01-18 at 10 53 35 AM
> 

It took me a second to notice this, but I think the screenshot you are referencing is the "before" portion of the video in the `main` branch - I switch it to this branches changes (at around 0:13 seconds) to show the adjustment - the annual records should be ordered by their starting month.

**Comment by morden35** [comment]
> Thank you so much for the review and feedback, Michelle!
> 
> > This is looking great! Would it be too much to also sort by month after we sort by year? For example, we can see in the screenshot below that if an agency has multiple different fiscal year reports, the months are not ordered. Preferably these reports would also be ordered in descending order. I think this is probably fine to do in a follow up pr if it's a heavy lift.
> > Screenshot 2024-01-18 at 10 53 35 AM
> 
> It took me a second to notice this, but I think the screenshot you are referencing is the "before" portion of the video in the `main` branch - I switch it to this branches changes (at around 0:13 seconds) to show the adjustment - the annual records should be ordered by their starting month. [It totally doesn't help that there's no commentary in the video because it totally looks like I'm showing the changes at the beginning - sorry about that! If there were commentary, the beginning would be me saying something like - "here take a look at what it looks like currently before I show you what these changes will make it look like".]

You are absolutely right! Fantastic! I guess you are already taking care of this in your new `sortReportsAnnualFirstDescending` function, right?

**Comment by mxosman** [comment]
> You are absolutely right! Fantastic! I guess you are already taking care of this in your new `sortReportsAnnualFirstDescending` function, right?

Spot on! I'll add comments to that function to make it more clear.

**Review by morden35 — APPROVED** [review]
lgtm!!! awesome stuff

**Pull request merged**